collect_app/src/main/java/org/odk/collect/android/tasks/GoogleSheetsAbstractUploader.java

```diff
@@ -231,31 +220,163 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
+
+                // check if Submissions folder exists
+                files = new ArrayList<>();
+                try {
+                    getFilesFromDrive(GOOGLE_DRIVE_SUBFOLDER, files, rootFolderId);
+                } catch (IOException e) {
+                    e.printStackTrace();
+                    mResults.put(id, form_fail + e.getMessage());
+                    return false;
+                }
+
+                folder = null;
+
+                for (com.google.api.services.drive.model.File file : files) {
+                    if (folder == null) {
+                        folder = file;
+                    } else {
+                        mResults.put(id, form_fail + files.size() + " folders found");
+                        return false;
+                    }
+                }
+
+                // if sub folder is not found then create a new one
+                if (folder == null) {
+                    try {
+                        folder = createFolderInDrive(GOOGLE_DRIVE_SUBFOLDER, rootFolderId);
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        mResults.put(id, form_fail + e.getMessage());
+                        return false;
+                    }
+                }
+
+                String subFolderId = folder.getId();
+
+                // check if jrformId folder exists
+                files = new ArrayList<>();
+                try {
+                    getFilesFromDrive(jrFormId, files, subFolderId);
+                } catch (IOException e) {
+                    e.printStackTrace();
+                    mResults.put(id, form_fail + e.getMessage());
+                    return false;
+                }
+
+                folder = null;
+
+                for (com.google.api.services.drive.model.File file : files) {
+                    if (folder == null) {
+                        folder = file;
+                    } else {
+                        mResults.put(id, form_fail + files.size() + " folders found");
+                        return false;
+                    }
+                }
+
+                // if sub folder is not found then create a new one
+                if (folder == null) {
+                    try {
+                        folder = createFolderInDrive(jrFormId, subFolderId);
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        mResults.put(id, form_fail + e.getMessage());
+                        return false;
+                    }
+                }
+
+                String uploadedFileId;
+
+                // file is ready to be uploaded
+                try {
+                    uploadedFileId = uploadFileToDrive(photosToUpload, key, folder, toUpload);
+                } catch (IOException e) {
+                    e.printStackTrace();
+                    mResults.put(id, form_fail + e.getMessage());
```

Comment:
I think this `form_fail` constant can also be removed. You can do that in a separate pull request if you'd prefer to do a separate cleanup pass for things you left the same as previous code.